drivers: pcie: endpoint: pcie_ep_iproc: use of undocumented / optional devicetree binding

**Describe the bug**
There is an unconditional reference to a `dmas` Devicetree property in `brcm_iproc_pcie_ep.c` but that property is undocumented in [brcm,iproc-pcie-ep.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/pcie/endpoint/brcm%2Ciproc-pcie-ep.yaml).

From what I can tell, this bug has been present since DMA support was added in 077522e6bbaadc92b1269efc021bc16f2b7ed241.

Unfortunately, because we lack a "build all" testsuite for pcie (or for that matter any kind of pcie testsuite), it has just gone unnoticed perpetually.

Please also mention any information which could help others to understand
the problem you're facing:
 - What target platform are you using? `bcm958402m2_m7` (with pl330 DMA disabled)
 - What have you tried to diagnose or workaround this issue? The solution is pretty straightforward - compile-out code conditionally where the `dmas` property does not exist.
 - Is this a regression? Yes, at 077522e6bbaadc92b1269efc021bc16f2b7ed241

**To Reproduce**
Steps to reproduce the behavior:
1. comment out lines 53 and 54 in [viper-common.dtsi](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm64/broadcom/viper-common.dtsi#L53)
2. `west build -b bcm958402m2_m7 samples/hello_world`
3. See error

**Expected behavior**
The pl330 `dmas` property is optional for Broadcom's customers, and so the upstream driver should compile without those properties. However, the `dmas` parameter should also be a documented (optional) feature in `brcm,iproc-pcie-ep.yaml`. If the feature is not present in the device tree, then it should not be compiled-in to the driver. Additionally, there should be a "build all" testsuite for pcie endpoints that ensures this driver can be built with various configurations.

**Impact**
Currently, it's a bit of a showstopper. We are engaged with Broadcom on this matter as well.

**Logs and console output**
```shell
ccache /Users/cfriedt/zephyr-sdk-0.15.2/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DKERNEL -D__PROGRAM_START -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I/Users/cfriedt/zephyrproject/zephyr/include -I/Users/cfriedt/zephyrproject/zephyr/build/zephyr/include/generated -I/Users/cfriedt/zephyrproject/zephyr/soc/arm/bcm_vk/viper -I/Users/cfriedt/zephyrproject/zephyr/soc/arm/bcm_vk/viper/. -I/Users/cfriedt/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include -isystem /Users/cfriedt/zephyrproject/zephyr/lib/libc/minimal/include -isystem /Users/cfriedt/zephyr-sdk-0.15.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/include -isystem /Users/cfriedt/zephyr-sdk-0.15.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/include-fixed -fno-strict-aliasing -Os -imacros /Users/cfriedt/zephyrproject/zephyr/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m7 -mthumb -mabi=aapcs --sysroot=/Users/cfriedt/zephyr-sdk-0.15.2/arm-zephyr-eabi/arm-zephyr-eabi -imacros /Users/cfriedt/zephyrproject/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/Users/cfriedt/zephyrproject/zephyr/samples/hello_world=CMAKE_SOURCE_DIR -fmacro-prefix-map=/Users/cfriedt/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/Users/cfriedt/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT zephyr/drivers/pcie/endpoint/CMakeFiles/drivers__pcie__endpoint.dir/pcie_ep_iproc.c.obj -MF zephyr/drivers/pcie/endpoint/CMakeFiles/drivers__pcie__endpoint.dir/pcie_ep_iproc.c.obj.d -o zephyr/drivers/pcie/endpoint/CMakeFiles/drivers__pcie__endpoint.dir/pcie_ep_iproc.c.obj -c /Users/cfriedt/zephyrproject/zephyr/drivers/pcie/endpoint/pcie_ep_iproc.c
In file included from /Users/cfriedt/zephyrproject/zephyr/include/zephyr/toolchain/gcc.h:92,
                 from /Users/cfriedt/zephyrproject/zephyr/include/zephyr/toolchain.h:50,
                 from /Users/cfriedt/zephyrproject/zephyr/include/zephyr/kernel_includes.h:19,
                 from /Users/cfriedt/zephyrproject/zephyr/include/zephyr/kernel.h:17,
                 from /Users/cfriedt/zephyrproject/zephyr/include/zephyr/drivers/dma.h:16,
                 from /Users/cfriedt/zephyrproject/zephyr/drivers/pcie/endpoint/pcie_ep_iproc.c:7:
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/device.h:84:41: error: '__device_dts_ord_DT_N_S_pcie_S_pcie_4e100000_P_dmas_IDX_0_PH_ORD' undeclared here (not in a function)
   84 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                         ^~~~~~~~~
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/toolchain/common.h:132:26: note: in definition of macro '_DO_CONCAT'
  132 | #define _DO_CONCAT(x, y) x ## y
      |                          ^
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/device.h:84:33: note: in expansion of macro '_CONCAT'
   84 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                 ^~~~~~~
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/device.h:210:37: note: in expansion of macro 'DEVICE_NAME_GET'
  210 | #define DEVICE_DT_NAME_GET(node_id) DEVICE_NAME_GET(Z_DEVICE_DT_DEV_ID(node_id))
      |                                     ^~~~~~~~~~~~~~~
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/device.h:227:34: note: in expansion of macro 'DEVICE_DT_NAME_GET'
  227 | #define DEVICE_DT_GET(node_id) (&DEVICE_DT_NAME_GET(node_id))
      |                                  ^~~~~~~~~~~~~~~~~~
/Users/cfriedt/zephyrproject/zephyr/drivers/pcie/endpoint/pcie_ep_iproc.c:480:22: note: in expansion of macro 'DEVICE_DT_GET'
  480 |         .pl330_dev = DEVICE_DT_GET(DT_INST_DMAS_CTLR_BY_IDX(0, 0)),
      |                      ^~~~~~~~~~~~~
In file included from /Users/cfriedt/zephyrproject/zephyr/include/zephyr/arch/arm/aarch32/arch.h:20,
                 from /Users/cfriedt/zephyrproject/zephyr/include/zephyr/arch/cpu.h:19,
                 from /Users/cfriedt/zephyrproject/zephyr/include/zephyr/kernel_includes.h:33:
/Users/cfriedt/zephyrproject/zephyr/build/zephyr/include/generated/devicetree_generated.h:740:40: error: 'DT_N_S_pcie_S_pcie_4e100000_P_dmas_NAME_txdma_VAL_channel' undeclared here (not in a function); did you mean 'DT_N_S_pcie_S_pcie_4e100000_IRQ_NAME_perst_VAL_irq'?
  740 | #define DT_N_INST_0_brcm_iproc_pcie_ep DT_N_S_pcie_S_pcie_4e100000
      |                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/devicetree.h:4107:9: note: in definition of macro 'DT_CAT7'
 4107 |         a1 ## a2 ## a3 ## a4 ## a5 ## a6 ## a7
      |         ^~
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/devicetree/dma.h:221:9: note: in expansion of macro 'DT_PHA_BY_NAME'
  221 |         DT_PHA_BY_NAME(node_id, dmas, name, cell)
      |         ^~~~~~~~~~~~~~
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/devicetree/dma.h:233:9: note: in expansion of macro 'DT_DMAS_CELL_BY_NAME'
  233 |         DT_DMAS_CELL_BY_NAME(DT_DRV_INST(inst), name, cell)
      |         ^~~~~~~~~~~~~~~~~~~~
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/sys/util_internal.h:105:36: note: in expansion of macro 'DT_N_INST_0_brcm_iproc_pcie_ep'
  105 | #define UTIL_PRIMITIVE_CAT(a, ...) a##__VA_ARGS__
      |                                    ^
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/sys/util_internal.h:104:26: note: in expansion of macro 'UTIL_PRIMITIVE_CAT'
  104 | #define UTIL_CAT(a, ...) UTIL_PRIMITIVE_CAT(a, __VA_ARGS__)
      |                          ^~~~~~~~~~~~~~~~~~
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/devicetree.h:328:31: note: in expansion of macro 'UTIL_CAT'
  328 | #define DT_INST(inst, compat) UTIL_CAT(DT_N_INST, DT_DASH(inst, compat))
      |                               ^~~~~~~~
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/devicetree.h:3154:27: note: in expansion of macro 'DT_INST'
 3154 | #define DT_DRV_INST(inst) DT_INST(inst, DT_DRV_COMPAT)
      |                           ^~~~~~~
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/devicetree/dma.h:233:30: note: in expansion of macro 'DT_DRV_INST'
  233 |         DT_DMAS_CELL_BY_NAME(DT_DRV_INST(inst), name, cell)
      |                              ^~~~~~~~~~~
/Users/cfriedt/zephyrproject/zephyr/drivers/pcie/endpoint/pcie_ep_iproc.c:481:29: note: in expansion of macro 'DT_INST_DMAS_CELL_BY_NAME'
  481 |         .pl330_tx_chan_id = DT_INST_DMAS_CELL_BY_NAME(0, txdma, channel),
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~~
/Users/cfriedt/zephyrproject/zephyr/build/zephyr/include/generated/devicetree_generated.h:740:40: error: 'DT_N_S_pcie_S_pcie_4e100000_P_dmas_NAME_rxdma_VAL_channel' undeclared here (not in a function); did you mean 'DT_N_S_pcie_S_pcie_4e100000_IRQ_NAME_perst_VAL_irq'?
  740 | #define DT_N_INST_0_brcm_iproc_pcie_ep DT_N_S_pcie_S_pcie_4e100000
      |                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/devicetree.h:4107:9: note: in definition of macro 'DT_CAT7'
 4107 |         a1 ## a2 ## a3 ## a4 ## a5 ## a6 ## a7
      |         ^~
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/devicetree/dma.h:221:9: note: in expansion of macro 'DT_PHA_BY_NAME'
  221 |         DT_PHA_BY_NAME(node_id, dmas, name, cell)
      |         ^~~~~~~~~~~~~~
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/devicetree/dma.h:233:9: note: in expansion of macro 'DT_DMAS_CELL_BY_NAME'
  233 |         DT_DMAS_CELL_BY_NAME(DT_DRV_INST(inst), name, cell)
      |         ^~~~~~~~~~~~~~~~~~~~
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/sys/util_internal.h:105:36: note: in expansion of macro 'DT_N_INST_0_brcm_iproc_pcie_ep'
  105 | #define UTIL_PRIMITIVE_CAT(a, ...) a##__VA_ARGS__
      |                                    ^
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/sys/util_internal.h:104:26: note: in expansion of macro 'UTIL_PRIMITIVE_CAT'
  104 | #define UTIL_CAT(a, ...) UTIL_PRIMITIVE_CAT(a, __VA_ARGS__)
      |                          ^~~~~~~~~~~~~~~~~~
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/devicetree.h:328:31: note: in expansion of macro 'UTIL_CAT'
  328 | #define DT_INST(inst, compat) UTIL_CAT(DT_N_INST, DT_DASH(inst, compat))
      |                               ^~~~~~~~
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/devicetree.h:3154:27: note: in expansion of macro 'DT_INST'
 3154 | #define DT_DRV_INST(inst) DT_INST(inst, DT_DRV_COMPAT)
      |                           ^~~~~~~
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/devicetree/dma.h:233:30: note: in expansion of macro 'DT_DRV_INST'
  233 |         DT_DMAS_CELL_BY_NAME(DT_DRV_INST(inst), name, cell)
      |                              ^~~~~~~~~~~
/Users/cfriedt/zephyrproject/zephyr/drivers/pcie/endpoint/pcie_ep_iproc.c:482:29: note: in expansion of macro 'DT_INST_DMAS_CELL_BY_NAME'
  482 |         .pl330_rx_chan_id = DT_INST_DMAS_CELL_BY_NAME(0, rxdma, channel),
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~~
[84/137] Building C object zephyr/lib/libc/minimal/CMakeFiles/lib__libc__minimal.dir/source/string/string.c.obj
ninja: build stopped: subcommand failed.
```

**Environment (please complete the following information):**
 - OS: Linux / macOS
 - Toolchain: Zephyr SDK 0.16.0
 - Commit SHA or Version used: d2917610a5ba2f190c7e8cfa49db93fab0e2f99e

**Additional context**
The `bcm958402m2_m7` is customizable for Broadcom's customers, and in some customizations, the pl330 dma is not present.

Technically, the `bcm958402m2_a72` platform would also have the same problem.

To prevent this kind of issue with PCIe Endpoint Drivers, we should have a build-all testsuite for them.

--- a/drivers/pcie/endpoint/pcie_ep_iproc.c
+++ b/drivers/pcie/endpoint/pcie_ep_iproc.c
@@ -469,7 +469,7 @@ err_out:
 
 static struct iproc_pcie_ep_ctx iproc_pcie_ep_ctx_0;
 
-static struct iproc_pcie_ep_config iproc_pcie_ep_config_0 = {
+static const struct iproc_pcie_ep_config iproc_pcie_ep_config_0 = {
 	.id = 0,
 	.base = (struct iproc_pcie_reg *)DT_INST_REG_ADDR(0),
 	.reg_size = DT_INST_REG_SIZE(0),
@@ -484,7 +484,7 @@ static struct iproc_pcie_ep_config iproc_pcie_ep_config_0 = {
 #endif
 };
 
-static struct pcie_ep_driver_api iproc_pcie_ep_api = {
+static const struct pcie_ep_driver_api iproc_pcie_ep_api = {
 	.conf_read = iproc_pcie_conf_read,
 	.conf_write = iproc_pcie_conf_write,
 	.map_addr = iproc_pcie_map_addr,

--- a/drivers/pcie/endpoint/pcie_ep_iproc.c
+++ b/drivers/pcie/endpoint/pcie_ep_iproc.c
@@ -477,9 +477,11 @@ static struct iproc_pcie_ep_config iproc_pcie_ep_config_0 = {
 	.map_low_size = DT_INST_REG_SIZE_BY_NAME(0, map_lowmem),
 	.map_high_base = DT_INST_REG_ADDR_BY_NAME(0, map_highmem),
 	.map_high_size = DT_INST_REG_SIZE_BY_NAME(0, map_highmem),
+#if DT_INST_NODE_HAS_PROP(0, dmas)
 	.pl330_dev = DEVICE_DT_GET(DT_INST_DMAS_CTLR_BY_IDX(0, 0)),
 	.pl330_tx_chan_id = DT_INST_DMAS_CELL_BY_NAME(0, txdma, channel),
 	.pl330_rx_chan_id = DT_INST_DMAS_CELL_BY_NAME(0, rxdma, channel),
+#endif
 };
 
 static struct pcie_ep_driver_api iproc_pcie_ep_api = {
@@ -489,7 +491,9 @@ static struct pcie_ep_driver_api iproc_pcie_ep_api = {
 	.unmap_addr = iproc_pcie_unmap_addr,
 	.raise_irq = iproc_pcie_raise_irq,
 	.register_reset_cb = iproc_pcie_register_reset_cb,
+#if DT_INST_NODE_HAS_PROP(0, dmas)
 	.dma_xfer = iproc_pcie_pl330_dma_xfer,
+#endif
 };
 
 DEVICE_DT_INST_DEFINE(0, &iproc_pcie_ep_init, NULL,

